python/python/bystro/supervised_ppca/_base.py

```diff
@@ -0,0 +1,516 @@
+import pytz
+from bystro._template_sgd_np import _BaseSGDModel # type: ignore
+
+
+class BaseGaussianFactorModel(_BaseSGDModel):
+    def __init__(self, n_components=2):
+        """
+        This is the base class of the model. Will never be called directly.
+
+        Parameters
+        ----------
+        n_components : int,default=2
+            The latent dimensionality
+
+        Sets
+        ----
+        creationDate : datetime
+            The date/time that the object was created
+        """
+        self.n_components = int(n_components)
+        self.creationDate = dt.now(pytz.timezone("US/Pacific"))
+
+    @abstractmethod
+    def fit(self, *args):
+        """
+        Fits a model given covariates X as well as option labels y in the
+        supervised methods
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,n_covariates)
+            The data
+
+        other arguments 
+
+        Returns
+        -------
+        self : object
+            The model
+        """
+
+    @abstractmethod
+    def get_covariance(self):
+        """
+        Gets the covariance matrix defined by the model parameters
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        covariance : np.array-like(p,p)
+            The covariance matrix
+        """
+
+    @abstractmethod
+    def _save_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of tensorflow variables saved
+        """
+
+    def get_precision(self):
+        """
+        Gets the precision matrix defined as the inverse of the covariance
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        precision : np.array-like(p,p)
+            The inverse of the covariance matrix
+        """
+        covariance = self.get_covariance()
+        precision = la.inv(covariance)
+        return precision
+
+    def get_stable_rank(self):
+        """
+        Returns the stable rank defined as
+        ||A||_F^2/||A||^2
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        srank : float
+            The stable rank. See Vershynin High dimensional probability for 
+            discussion, but this is a statistically stable approximation to rank
+        """
+        covariance = self.get_covariance()
+        srank = _get_stable_rank(covariance)
+        return srank
+
+    def transform(self, X):
+        """
+        This returns the latent variable estimates given X
+
+        Parameters
+        ----------
+        X : np array-like,(N_samples,\sum idxs)
+            The data to transform
+
+        Returns
+        -------
+        S : np.array-like,(N_samples,n_components)
+            The factor estimates
+        """
+        prec = self.get_precision()
+        coefs = np.dot(self.W_, prec)
```

Comment:
Grew up old school before @ existed. The *very tiny* advantage of dot is that it's easier to control the order of evaluations of multiplications which can save time depending on matrix shapes. Do we want to make @ a standard?

Reply:
I def agree that we sometimes have to be careful about how to order matrix multiplications in order to avoid storing large intermediate values.  For example, if we have four matrices A (N x N), B (N x N) and C (N x N) and D (N x m) with N >> m, then we'll want to be sure to compute the matrix product ABCD as (A (B (C D))) in order to avoid directly computing AB or BC.

With the matrix mult operator that looks like:

```python
 (A @ (B @ (C @ D)))
```

and with np.dot it looks like:
```python
 np.dot(A, np.dot(B, np.dot(C, D)))
```

So we get the same control, but the matrix mult operator is (imho) easier to both read and type, YMMV :)

Reply:
Will possibly change in the future if that's OK.